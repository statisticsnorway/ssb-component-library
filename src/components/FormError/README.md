FormError
========

> Dialog box for form errors

### Usage

#### HTML

```html
<div class="ssb-form-error">
    <span class="error-icon"></span>
    <div>
        <span class="error-title">Some fields are not right!</span>
        <ul>
            <li>error 1</li>
            <li>error 2</li>
            <li>error 3</li>
        </ul>
    </div>
</div>

<div class="ssb-form-error negative">
    <span class="error-icon"></span>
    <div>
        <span class="error-title">Some fields are not right!</span>
        <ul>
            <li>error 1</li>
            <li>error 2</li>
            <li>error 3</li>
        </ul>
    </div>
</div>
```

#### React

```jsx harmony
const errors = ['error 1', 'error 2', 'error 3'];

<FormError errorMessages={errors} title="Some fields are not right:" />

<FormError negative errorMessages={errors} title="Some fields are not right!" />
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| className | string | Optional container class |
| errorMessages | array | Renders list items |
| negative | Optional boolean | Will change component style. Defaults to false |
| title | string | Renders a title |
