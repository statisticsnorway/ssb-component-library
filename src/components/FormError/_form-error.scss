.ssb-form-error {
  @include roboto;
  background: $ssb-red-1;
  border-left: 4px solid $ssb-red-4;
  display: flex;
  padding: 24px 24px 24px 0;
  width: 100%;

  .error-icon {
    margin: 0 15px;

    &:before {
      align-items: center;
      background: $ssb-red-4;
      border-radius: 2px;
      color: $ssb-white;
      content: "!";
      display: inline-flex;
      font-size: 18px;
      font-weight: bold;
      height: 26px;
      justify-content: center;
      width: 26px;
    }
  }

  .error-title {
    color: $ssb-red-4;
    font-weight: bold;
  }

  ul {
    margin-bottom: 0;
    margin-top: 15px;
    padding-left: 20px;
  }
}
