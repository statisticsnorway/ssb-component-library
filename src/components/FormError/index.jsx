import React from 'react';
import PropTypes from 'prop-types';

const FormError = ({ className, errorMessages, title }) => (
	<div className={`ssb-form-error ${className}`}>
		<span className="error-icon" />
		<div>
			<span className="error-title">{title}</span>
			<ul>
				{errorMessages.map((error, index) => (
					<li key={index}> {/* eslint-disable-line */}
						{error}
					</li>
				))}
			</ul>
		</div>
	</div>
);

FormError.defaultProps = {
	className: '',
	title: '',
};

FormError.propTypes = {
	className: PropTypes.string,
	errorMessages: PropTypes.array.isRequired,
	title: PropTypes.string,
};

export default FormError;
