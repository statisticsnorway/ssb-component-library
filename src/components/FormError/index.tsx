import React from 'react'
import classNames from '../../utils/utils'

interface FormErrorProps {
  className?: string
  errorMessages: string[]
  negative?: boolean
  title?: string
}

const FormError: React.FC<FormErrorProps> = ({ className, errorMessages, negative, title = '' }) => (
  <div className={classNames('ssb-form-error', negative && ' negative', className)}>
    <span className='error-icon' />
    <div>
      <span className='error-title'>{title}</span>
      <ul>
        {errorMessages.map((error, index) => (
          // eslint-disable-next-line react/no-array-index-key
          <li key={index}>{error}</li>
        ))}
      </ul>
    </div>
  </div>
)

export default FormError
