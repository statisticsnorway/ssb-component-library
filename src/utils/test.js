import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
