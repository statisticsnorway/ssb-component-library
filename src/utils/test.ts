import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
