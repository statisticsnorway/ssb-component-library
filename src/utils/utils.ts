// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
