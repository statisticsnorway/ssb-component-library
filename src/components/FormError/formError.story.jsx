import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import FormError from './index';

const errorMessages = [
	'Not a valid email',
	'Select a country',
	'Feed the dog',
];

storiesOf('FormError', module).addDecorator(centered)
	.add('Default', () => (
		<div style={{ width: '500px' }}>
			<FormError errorMessages={errorMessages} title="Some fields are not right!" />
		</div>
	));
