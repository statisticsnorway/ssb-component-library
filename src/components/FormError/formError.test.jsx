import React from 'react'
import { render } from '../../utils/test'
import FormError from './index'

const errorMessages = ['Not a valid email', 'Select a country', 'Feed the dog']

describe('FormError component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<FormError errorMessages={errorMessages} title='Some fields are not right!' />)
    expect(asFragment()).toMatchSnapshot()
  })
})
