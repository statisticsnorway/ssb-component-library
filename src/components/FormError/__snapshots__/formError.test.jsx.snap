// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormError component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-form-error"
  >
    <span
      class="error-icon"
    />
    <div>
      <span
        class="error-title"
      >
        Some fields are not right!
      </span>
      <ul>
        <li>
          Not a valid email
        </li>
        <li>
          Select a country
        </li>
        <li>
          Feed the dog
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
